doc/src/documentation_style_guide.rst

```diff
@@ -0,0 +1,1019 @@
+<http://docutils.sourceforge.net/rst.html>`_ format extended by `Sphinx
+<http://www.sphinx-doc.org/en/master/>`_. Here is a concise guide to `Quick
+reStructuredText <http://docutils.sourceforge.net/docs/user/rst/quickref.html>`_. More in-depth
+information about using reStructuredText can be found in the `Sphinx
+Documentation
+<http://www.sphinx-doc.org/en/master/usage/restructuredtext/index.html>`_.
+
+In order for Sphinx to render docstrings nicely in the HTML documentation, some
+formatting guidelines should be followed when writing docstrings:
+
+* Always use """triple double quotes""" around docstrings. Use r"""raw triple
+  double quotes""" if you use any backslashes in your docstrings.
+* Include a blank line before the docstring’s closing quotes.
+* Lines should not be longer than 80 characters.
+* Always write class-level docstrings under the class definition line, as that
+  is better to read in the source code.
+* The various methods on the class can be mentioned in the docstring or
+  examples if they are important, but details on them should go in the
+  docstring for the method itself.
+* In order to make section underlining work nicely in docstrings, `numpydoc
+  Sphinx extension <https://pypi.org/project/numpydoc/>`_ is used. (WORK IN
+  PROGRESS: this may change via issues `#17599
+  <https://github.com/sympy/sympy/issues/17599>`_ and `#17618
+  <https://github.com/sympy/sympy/issues/17618>`_.)
+* Always double check that you have formatted your docstring correctly:
+
+1. Make sure that your docstring is imported into Sphinx.
+2. Build the Sphinx docs (``cd doc; make html``).
+3. Make sure that Sphinx doesn't output any errors.
+4. Open the page in ``_build/html`` and make sure that it is formatted
+   correctly.
+
+Docstring Sections
+------------------
+
+In SymPy’s docstrings, it is preferred that function, class, and method
+docstrings consist of the following sections in this order:
+
+1. Single-Sentence Summary
+2. Explanation
+3. Examples
+4. Parameters
+5. See Also
+6. References
+
+WORK IN PROGRESS via issue `#17643
+<https://github.com/sympy/sympy/issues/17643>`_: The Single-Sentence Summary and
+Examples sections are **required** for every docstring. Checks will not pass if
+these sections are not included in a docstring.
+
+Do not change the names of these supported sections, for example, the heading
+“Examples” as a plural should be used even if there is only one example. If you
+wish to include a section that is not listed in these main section headings, you
+can add a subheading of your choice under Explanation, Examples, or Parameters.
+(WORK IN PROGRESS via issue `#17618
+<https://github.com/sympy/sympy/issues/17618>`_.)
+
+Additionally, SymPy will continue to support all of the section headings listed
+in the `NumPy Docstring Guide
+<https://numpydoc.readthedocs.io/en/latest/format.html>`_.
+
+Headings should be underlined with the same length in equals signs and
```

Comment:
Are you sure the headings should be underlined with equals signs?

NumPy seems to use hyphens: https://numpydoc.readthedocs.io/en/latest/format.html#sections

Please note that if this is changed to hyphens, `pydocstyle` will complain about the following blank line:

    D412: No blank lines allowed between a section header and its content ('Examples')

Reply:
Presently it doesn't matter what you use. The plan was to potentially change this (https://github.com/sympy/sympy/issues/17618). For now, equals signs are used almost unilaterally in the SymPy codebase, so this guide reflects that.

Reply:
> For now, equals signs are used almost unilaterally in the SymPy codebase, so this guide reflects that.

OK, but I thought we are actually in the process of changing some of the guidelines.

I thought the idea was:

1. write a style guide
2. change all docstrings to conform

Embracing the NumPy style seems like a good idea to me.
I wouldn't suggest this at any other time, but since all docstrings will be rewritten anyway, this seems like a good opportunity.

Reply:
> OK, but I thought we are actually in the process of changing some of the guidelines.

I'd like to see this happen. We have a chance to change things that are not preferable if if widely used in SymPy.

Reply:
I think it makes sense to use equals signs under these section headings because these are truly the different sections of each docstring, and equals signs are most commonly used as the next heading level down from the title (with overline), which also goes with the Python recommended heading conventions that was suggested that I add to the formatting section. And if we use equals signs for the section, we can use hyphens for subheadings if we are able to add that functionality.

The idea of the GSoD project is to create a style guide with preferred formatting and edit some docstrings to reflect it to serve as a model, but it's not feasible within the scope of the project to rewrite all of the docstrings. If SymPy wants to switch to hyphens for docstring section underlining to be more similar to other projects, we can put that in the style guide and have it as a goal for the future, but that work wouldn't happen within the scope of GSoD.

Reply:
> I think it makes sense to use equals signs under these section headings because these are truly the different sections of each docstring

I disagree.
I would consider the whole docstring as one (implicit) section, having a few sub-sections.

> And if we use equals signs for the section, we can use hyphens for subheadings if we are able to add that functionality.

Can you clarify what you mean by "add that functionality"?

The `numpydoc` extension apparently doesn't support subheadings. It treats `====` and `----` underlines equally.
I don't think that a further level of headings is necessary.

Reply:
I meant this PR to change the numpydoc extension to support subheadings in the docstrings: https://github.com/sympy/sympy/issues/17618. The idea behind that being that if someone wants to add a subsection/piece of information to their docstring that doesn't fit under the supported headings, they could add a subheading. But this is still a WIP and may not happen before we need to merge the guide for GSoD.

What do others think about changing the current ==== underlining in the docstrings to -----? This would be a work in progress that would extend beyond GSoD, but if others also want the docstrings to look more similar to other projects, now would be the time to put that in the style guide as a goal for the future.

Reply:
I meant this PR to change the numpydoc extension to support subheadings in the docstrings: #17618. The idea behind that being that if someone wants to add a subsection/piece of information to their docstring that doesn't fit under the supported headings, they could add a subheading. But this is still a WIP and may not happen before we need to merge the guide for GSoD.

What do others think about changing the current ==== underlining in the docstrings to -----? This would be a work in progress that would extend beyond GSoD, but if others also want the docstrings to look more similar to other projects, now would be the time to put that in the style guide as a goal for the future.

Reply:
My preference would be to keep the equals signs, as we are currently doing throughout SymPy.